### ISSUE TITLE
PLIO TypeError When Trying to Write Out a v5 Control Network

### ISSUE THREAD

**Issue by astrostu** [USGS-Astrogeology/plio#141]
Hi:  A couple things in PLIO have been improved and fixed since I last tried this 9 months ago, so I wanted to try again to write out a control network (plio-1.2.2-3_dev).  I read in a version 5 control network binary, and then try to write it out using io_controlnetwork.to_isis() and explicitly setting version=5.

The error I get is:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stu/anaconda3/envs/plio/lib/python3.7/site-packages/plio/io/io_controlnetwork.py", line 60, in to_isis
    point_messages, point_sizes = store.create_points(obj, pointid_prefix, pointid_suffix)
  File "/Users/stu/anaconda3/envs/plio/lib/python3.7/site-packages/plio/io/io_controlnetwork.py", line 292, in create_points
    getattr(measure_spec, attr).extend(m[df_attr])
TypeError: Parameter to MergeFrom() must be instance of same class: expected Isis.ControlPointFileEntryV0002.Measure.MeasureLogData got Isis.ControlPointFileEntryV0005.Measure.MeasureLogData.
```

I get this also if I try to explicitly write it out as a v2 network, which I tried as a "let's see what happens."

I don't see any text "MergeFrom" in io_controlnetwork.py.

Searching around, I found some site that suggested that Python v 3.7 might have a bug with this sort of assignment.  In my PLIO conda environment, I tried running an update on Python and went from 3.7.3 to 3.8.2 (and it upgraded to plio-1.2.3-py38_0).  Repeated, and now I get a different error:
```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/stu/anaconda3/envs/plio/lib/python3.8/site-packages/plio/io/io_controlnetwork.py", line 60, in to_isis
    point_messages, point_sizes = store.create_points(obj, pointid_prefix, pointid_suffix)
  File "/Users/stu/anaconda3/envs/plio/lib/python3.8/site-packages/plio/io/io_controlnetwork.py", line 274, in create_points
    setattr(point_spec, attr, attrtype(g.iloc[0][df_attr]))
AttributeError: Assignment not allowed to repeated field "log" in protocol message object.
```

With the fix to PLIO a few months ago, there are no two columns named "log" (it's pointLog and measureLog), so I still don't know what the error is, other than I suspect it's a control net v5 write issue.

I will note as before that line 254 of the io_controlnetwork.py file explicitly defines point_spec = cnf.ControlPointFileEntryV0002() and there is no switch for v5.  I apparently don't know enough about how these files are set up to try to fix it: When I change it to point_spec = cnp5.ControlPointFileEntryV0005_pb2(), I get this error:

`AttributeError: module 'plio.io.ControlPointFileEntryV0005_pb2' has no attribute 'ControlPointFileEntryV0005_pb2'`

I got that also before I upgraded to the new version of Python when I tried to change that line in io_controlnetwork.py.

I'll also note that ControlPointFileEntryV0005_pb2.py also has a .log attribute in the structure (not sure of the correct coding term), not pointLog and measureLog.

I appreciate your unfunded work on this, perhaps these notes will help diagnose and fix things.

**Comment by jessemapel** [comment]
This looks like it's a bug in PLIO and not somewhere else. What's happening is that your read-in control network has V5 components in it and then when it tries to write it out, it attempts to append them to V2 components which results in that error.

The correct fix here is to properly un-pack the control measure log data and control point log data into Python structures and then re-pack them when writing out. For right now, try just assigning an empty list to both the measureLog and pointLog columns in your dataframe. Those are only used for tracking changes to the control network and should not affect any processing.

**Comment by astrostu** [comment]
Storing blanks to the pointLog and measureLog columns results in the same error.  Deleting them lets me write out a network.  But, ISIS3 then can't read it:

```
**I/O ERROR** Reading the control network [test.net] failed.
**I/O ERROR** Failed to convert protobuf version 2 control point at index [2] into a ControlPoint.
**PROGRAMMER ERROR** The SerialNumber is not unique. A measure with serial number [MRO/CTX/1073920639:230] already exists for ControlPoint [TEST_99874521_ptG00000001].
```

It also output a 2.0 MB control network, 10% larger than the 1.8 MB network it had read in.  My naïve interpretation is that it's still trying to write out some amalgamation of a v2 and v5 network.

This is what I as trying to do to write it out:
`io_controlnetwork.to_isis(df, 'test.net', mode='wb', version=5, headerstartbyte=65536, targetname='Mars')
`

Also, I know that your suggestion was just a temp fix to see if it would work, but I actually do use the information in measureLog because it contains the GoodnessOfFit data, which I then use for various purposes like my own implementation of thinning.  So eventually it would be good to keep.

**Comment by jessemapel** [comment]
That error looks like it's not related to the log data. Can you check that there isn't a repeated measure in that point?

**Comment by astrostu** [comment]
That point is comprised of two, unique measures.  And image MRO/CTX/1073920639:230 is not one of them.  
[cnet_master_FINALAUTOMATED_jigsaw.net.zip](https://github.com/USGS-Astrogeology/plio/files/4308732/cnet_master_FINALAUTOMATED_jigsaw.net.zip)

**Comment by astrostu** [comment]
P.S.  Here are the commands I'm running, from the command line.  Just trying to get a simple in-out working at this point:

```
import pandas as pd
from plio.io  import io_controlnetwork
myCNET = "cnet_master_FINALAUTOMATED_jigsaw.net"
df = io_controlnetwork.from_isis(myCNET)
del df['pointLog']
del df['measureLog']
io_controlnetwork.to_isis(df, 'test.net', mode='wb', version=5, headerstartbyte=65536, targetname='Mars')
```

**Comment by jessemapel** [comment]
This should be fixed by #146

**Comment by jlaura** [comment]
Sort of fixed. Plio only ever writes out v2 nets right now. But #146 fixes the problem for V2 networks and then ISIS will handle versioning.

**Comment by astrostu** [comment]
What is the difference between v2 and v5 control nets?  I haven't been able to find this documented anywhere.  I'm wondering if it's worth trying to do things with v2 so I can interact better with PLIO, or if I should hold off.

**Comment by jessemapel** [comment]
@astrostu  the difference has to do with how the data is stored in the file. The V2 networks have problems when they get very large (read multiple GB) that the V5 networks solve.

**Comment by jlaura** [comment]
Fixed by #146.
